File: 2020/day10.rb
def prompt
  r = readline()
  exit if r == "x\n"
end

test = false #true #
if test
  input = [1, 4, 5, 6, 7, 10, 11, 12, 15, 16, 19].insert(0, 0)
  input = File.read('day10.tst').split("\n").map(&:to_i).sort!.insert(0, 0)
  input.push(input[-1]+3)
else
  input = File.read('day10.txt').split("\n").map(&:to_i).sort!.insert(0, 0)
  input.push(input[-1]+3)
end

# pp input

puts "Part 1:"
puts (0...(input.length-1)).each
                            .map { |a| input[a+1]-input[a] }
                            .group_by { |n| n }
                            .map { |e, v| v.length }
                            .reduce(:*)

#  ____            _     ____  
# |  _ \ __ _ _ __| |_  |___ \ 
# | |_) / _` | '__| __|   __) |
# |  __/ (_| | |  | |_   / __/ 
# |_|   \__,_|_|   \__| |_____|

routes = {}
routes[0] = 1
input.each { |a|
  next if a == 0
  # puts("a:, #{a}")
  routes[a] = 0
  routes[a] += routes[a-1] if !routes[a-1].nil?
  routes[a] += routes[a-2] if !routes[a-2].nil?
  routes[a] += routes[a-3] if !routes[a-3].nil?
  # puts "routes: #{a} = #{routes[a]}"
  # routes[a] = routes[a-1] + routes[a-2] + routes[-3]
}

puts "Part 2: #{routes.max}"

# Starting at index 1 find how many adapters can be connected in sequence

File: 2020/day10.py
# keywords: zip, collection, map
import collections

input = open('day10.tst', 'r').readlines()
input = [1, 4, 5, 6, 7, 10, 11, 12, 15, 16, 19]

js = sorted(map(int, input))
js = [0] + js + [js[-1] + 3]

# for j1, j2 in zip(js, js[1:]):
#   print("j1:",j1,"j2:",j2,"j2-j1",j2-j1)

cnt = collections.Counter(j2 - j1 for j1, j2 in zip(js, js[1:]))
# print("Part 1:", cnt[1] * cnt[3])

dp = [1] + [0] * (len(js) - 1)
print("js:",js)
print("dp:",dp)
print()

for i, j in enumerate(js):
  print("i:",i,"j:",j)

  print(range(i - 3, i))
  exit()

  for l in range(i - 3, i):
    print("  l:",l,"js[l]:",js[l])

    if j - js[l] <= 3:
        dp[i] += dp[l]
    print(dp)
    print()

exit()

print(dp[-1])
